client/app/services/parameters/DateRangeParameter.js

```diff
@@ -0,0 +1,163 @@
+    value: () => [moment().subtract(1, 'month').startOf('month'), moment().subtract(1, 'month').endOf('month')],
+  },
+  last_year: {
+    name: 'Last year',
+    value: () => [moment().subtract(1, 'year').startOf('year'), moment().subtract(1, 'year').endOf('year')],
+  },
+  last_7_days: {
+    name: 'Last 7 days',
+    value: () => [moment().subtract(7, 'days'), moment()],
+  },
+};
+
+export const DynamicDateRangeType = PropTypes.oneOf(values(DYNAMIC_DATE_RANGES));
+
+export function isDynamicDateRangeString(value) {
+  if (!startsWith(value, DYNAMIC_PREFIX)) {
+    return false;
+  }
+  return !!DYNAMIC_DATE_RANGES[value.substring(DYNAMIC_PREFIX.length)];
+}
+
+export function isDynamicDateRange(value) {
+  return includes(DYNAMIC_DATE_RANGES, value);
+}
+
+export function getDynamicDateRangeFromString(value) {
+  if (!isDynamicDateRangeString(value)) {
+    return null;
+  }
+  return DYNAMIC_DATE_RANGES[value.substring(DYNAMIC_PREFIX.length)];
+}
+
+class DateRangeParameter extends Parameter {
+  constructor(parameter, parentQueryId) {
+    super(parameter, parentQueryId);
+    this.setValue(parameter.value);
+  }
+
+  get hasDynamicValue() {
+    return isDynamicDateRange(this.normalizedValue);
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  normalizeValue(value) {
+    if (isDynamicDateRangeString(value)) {
+      return getDynamicDateRangeFromString(value);
+    }
+
+    if (isDynamicDateRange(value)) {
+      return value;
+    }
+
+    if (isObject(value) && !isArray(value)) {
+      value = [value.start, value.end];
+    }
+
+    if (isArray(value) && (value.length === 2)) {
+      value = [moment(value[0]), moment(value[1])];
+      if (value[0].isValid() && value[1].isValid()) {
+        return value;
+      }
+    }
+    return null;
+  }
+
+  setValue(value) {
+    const normalizedValue = this.normalizeValue(value);
+    if (isDynamicDateRange(normalizedValue)) {
+      this.value = DYNAMIC_PREFIX + findKey(DYNAMIC_DATE_RANGES, normalizedValue);
+    } else if (isArray(normalizedValue)) {
+      this.value = {
+        start: normalizedValue[0].format(DATETIME_FORMATS[this.type]),
+        end: normalizedValue[1].format(DATETIME_FORMATS[this.type]),
+      };
+    } else {
+      this.value = normalizedValue;
+    }
+    this.$$value = normalizedValue;
+
+    this.updateLocals();
+    this.clearPendingValue();
+    return this;
+  }
+
+  getExecutionValue() {
+    if (this.hasDynamicValue) {
+      const format = date => date.format(DATETIME_FORMATS[this.type]);
+      const [start, end] = this.normalizedValue.value().map(format);
+      return { start, end };
+    }
+    return this.value;
+  }
+
+  toUrlParams() {
+    const prefix = this.urlPrefix;
+    if (isObject(this.value) && this.value.start && this.value.end) {
+      return {
+        [`${prefix}${this.name}`]: null,
+        [`${prefix}${this.name}.start`]: this.value.start,
+        [`${prefix}${this.name}.end`]: this.value.end,
```

Comment:
This `start`/`end` implementation affects the superclass and subclasses by forcing them to be aware of reseting these. Personally, I think that's not good practice cause it doesn't adhere to agnosticism, therefore making code maintenance tricky and complex.

Any way around it?

Maybe move these keys to the value, allowing all classes to implement nothing beyond `[${prefix}${this.name}]`?
```js
toUrlParams() {
    const prefix = this.urlPrefix;
    if (isObject(this.value) && this.value.start && this.value.end) {
      return {
        [`${prefix}${this.name}`]: `start::${this.value.start}--end::${this.value.end}`,
      }
  }
}
```

Reply:
I totally agree with you on this one, a different option is to use a different logic to remove unused parameters from the url. I'll explore those alternatives :slightly_smiling_face:

Reply:
I like the idea of going even simpler on this, like using only a small separator `p_date_range=[start]--[end]`. Beside gaining more organization on code we also reduce the size of the url :)

On the other hand, one downside of changing is for people that have saved url links with parameter values. It doesn't seem a big issue, but if it turns out to be, it's possible to support both url syntaxes.

Reply:
Yeah I think it would be easy and to add a short backwards compatibility code 👍